examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,39 +18,152 @@
+    {
+        size_t size    = 0;
+        CHIP_ERROR err = responses.Value().ComputeSize(&size);
+        VerifyOrExit(err == CHIP_NO_ERROR, ChipLogProgress(Controller, "HandlePresentMessagesRequest size check failed"));
+
+        MessageResponseOption * optionArray = new MessageResponseOption[size];
+        VerifyOrExit(optionArray != nullptr,
+                     ChipLogProgress(Controller, "HandlePresentMessagesRequest MessageResponseOption alloc failed"));
+        optionToFree.push_back(optionArray);
+
+        int counter = 0;
+        auto iter   = responses.Value().begin();
+        while (iter.Next())
+        {
+            auto & response = iter.GetValue();
+
+            ChipLogProgress(Controller, "HandlePresentMessagesRequest option id:%u", response.messageResponseID.ValueOr(0));
+
+            if (response.messageResponseID.HasValue())
+            {
+                optionArray[counter].messageResponseID = Optional<uint32_t>(response.messageResponseID.ValueOr(0));
+            }
+
+            if (response.label.HasValue())
+            {
+                char * labelBuffer = new char[response.label.Value().size()];
+                VerifyOrExit(labelBuffer != nullptr,
+                             ChipLogProgress(Controller, "HandlePresentMessagesRequest label alloc failed"));
+                needToFree.push_back(labelBuffer);
+                memcpy(labelBuffer, response.label.Value().data(), response.label.Value().size());
+
+                ChipLogProgress(Controller, "HandlePresentMessagesRequest option label:%s size:%lu", labelBuffer,
+                                response.label.Value().size());
+
+                optionArray[counter].label = Optional<CharSpan>(CharSpan::fromCharString(labelBuffer));
+            }
+            counter++;
+        }
+
+        Message message{ ByteSpan(messageIdBuffer, messageId.size()),
+                         priority,
+                         messageControl,
+                         startTime,
+                         duration,
+                         CharSpan::fromCharString(messageTextBuffer),
+                         Optional<DataModel::List<MessageResponseOption>>(
+                             DataModel::List<MessageResponseOption>(optionArray, size)) };
+
+        mMessages.push_back(message);
+    }
+    else
+    {
+        Message message{ ByteSpan(messageIdBuffer, messageId.size()), priority, messageControl, startTime, duration,
+                         CharSpan::fromCharString(messageTextBuffer) };
+        mMessages.push_back(message);
+    }
+
     // Add your code to present Message
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest complete");
+
+    return;
+exit:
+
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest error exit");
+    // only reach here on errors
+    if (messageIdBuffer != nullptr)
+    {
+        delete[] messageIdBuffer;
+    }
+    for (MessageResponseOption * memloc : optionToFree)
```

Comment:
There's only one in practice, right?

This generally seems pretty fragile; is there a reason to not just use owning RAII types for these pointers that will need to be deallocated, with `release()` calls happening as needed if the pointer should _not_ be deallocated?

Reply:
that is a refactoring to consider

Reply:
Refactored memory management into destructor of CachedMessage